### PR TITLE
prometheus: 2.27.1 -> 2.30.3

### PR DESCRIPTION
**Pull request by eyJhb**
###### Motivation for this change
Needs update, there are some syntax shown on ie. blackbox-exporter that does not work on the current version of prometheus (I think).

###### Things done
I have updated the version numbers + hashes.
However, they have changed yarn out with npm, which has caused some headache.
They have also added a module system, that is currently working OK with a hack that needs to be changed...
 
I have ran the prometheus tests and they pass without any issues.

TODOs

- [x] Cleanup code, ie. remove unused packages, buildInputs, etc.
- [x] Update webui update script

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by Ma27 — COMMENTED** [review]
Are there new releases of exporters that are broken according to you (e.g. `blackbox-exporter`)? if yes, these should be updated in here as well.

File: pkgs/servers/monitoring/prometheus/default.nix [hunk 1]
```diff
@@ -9,27 +12,59 @@
...
     owner = "prometheus";
     repo = "prometheus";
-    sha256 = "0836ygyvld5skjycd7366i6vyf451s6cay5ng6c2fwq0skvp2gj2";
+    sha256 = "1as6x5bsp7mxa4rp7jhyjlpcvzqm1zngnwvp73rc4rwhz8w8vm3k";
   };
 
   goPackagePath = "github.com/prometheus/prometheus";
 
-  webui = mkYarnPackage {
+  codemirrorNode = import ./webui/codemirror-promql {
```
[the remark this answers is not in the document]
Comment (eyJhb) [reply pkgs/servers/monitoring/prometheus/default.nix:26]: I don't think we should stick with yarn, but move to npm.
Looked at this thread here - https://github.com/prometheus/prometheus/pull/9198 which also links this https://github.com/thanos-io/thanos/issues/4562 and it seems to be a general move.

There are some dependencies yarn can't handle as well, that npm handles well. So we could end up having a different set of dependencies than upstream, if we continue to use yarn.

**Comment by basvandijk** [comment]
@GrahamcOfBorg test prometheus

File: pkgs/servers/monitoring/prometheus/default.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (basvandijk) [reply pkgs/servers/monitoring/prometheus/default.nix:26]: Agreed. If upstream is moving to npm we should do so as well.

**Review by basvandijk — APPROVED** [review]
LGTM!

**Comment by basvandijk** [comment]
@eyJhb I'm trying to upgrade to the latest prometheus-2.31.0 and I noticed they moved the [web/ui](https://github.com/prometheus/prometheus/tree/v2.31.0/web/ui) package to npm workspaces. Any idea how to update your `pkgs/servers/monitoring/prometheus/webui/update-web-deps.sh` to this new way?

**Comment by eyJhb** [comment]
> @eyJhb I'm [trying to upgrade](https://github.com/eyJhb/nixpkgs/compare/prometheus-2.3.3...basvandijk:prometheus-2.31.0) to the latest prometheus-2.31.0 and I noticed they moved the [web/ui](https://github.com/prometheus/prometheus/tree/v2.31.0/web/ui) package to npm workspaces. Any idea how to update your `pkgs/servers/monitoring/prometheus/webui/update-web-deps.sh` to this new way?

It just seems to be sugar on top of npm, and nothing more. Currently it just builds the two components independently, and then fixes the paths. 

Should that be any issue? The webui works fine when I tested it :) - Are you available on Matrix for discussing any miscommunication? :)

**Comment by basvandijk** [comment]
> It just seems to be sugar on top of npm, and nothing more. Currently it just builds the two components independently, and then fixes the paths.

The issue is that they now have a [single `package-lock.json` file in `web/ui`](https://github.com/prometheus/prometheus/blob/v2.31.0/web/ui/package-lock.json) instead of two separate ones in `web/ui/module/codemirror-promql` and `web/ui/react-app/`.

> Should that be any issue? The webui works fine when I tested it :)

At the moment the package doesn't build. But maybe I can try just using that single `package-lock.json` for the two packages.

> Are you available on Matrix for discussing any miscommunication? :)

I'm not and TIL about Matrix. I'll look into signing up. Where can I do that?

**Comment by basvandijk** [comment]
It seems this is ready to merge. We can do the 2.31.0 upgrade in another PR. Merging in now...

**Comment by eyJhb** [comment]
> It seems this is ready to merge. We can do the 2.31.0 upgrade in another PR. Merging in now...

Oh, didn't notice a new version had come out. Sorry!
Hopefully it can be fixed soon. Seems annoying that Prometheus cannot settle on anything, and keeps changing it around.
But what can one do ... :)

Thanks for merging @basvandijk ! :)

**Pull request merged**